### PR TITLE
RheinMainTV

### PR DESCRIPTION
**Pull request by barthelmannk**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Added support for rheinmaintv.de, that is
- added file yt_dlp/extractor/rheinmaintv.py and
- added a corresponding import in yt_dlp/extractor/_extractors.py.

Suggest to fix a potential problem in yt_dlp/extractor/common.py:
_search_regex can return a tuple, and _html_search_regex should be prepared for that.

The former (= support) does not depend on the latter (= suggested fix).

Fixes #

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/rheinmaintv.py [hunk 1]
```diff
@@ -0,0 +1,601 @@
+from ..utils import extract_attributes, float_or_none, int_or_none, parse_iso8601
+from .common import InfoExtractor
+
+
+class RheinMainTVIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?rheinmaintv\.de/sendungen/(?:[a-z-]+/)*(?P<display_id>[a-z-]+)/vom-(?P<date>[0-9]{2}\.[0-9]{2}\.[0-9]{4})(?:/(?P<serial_number>[0-9]+))?'
+    _TESTS = [{
+        'url': 'https://www.rheinmaintv.de/sendungen/beitrag-video/formationsgemeinschaft-rhein-main-bei-den-deutschen-meisterschaften/vom-14.11.2022/',
+        # 'md5': ...,  # left out because checksum is changing
+        'info_dict': {
```
Comment (pukkandan) [inline yt_dlp/extractor/rheinmaintv.py:10]: this is not how you write test. See other extractors. Or, delete all the keys except ext/id and run the test. The error message will tell you most of the fields needed

File: yt_dlp/extractor/rheinmaintv.py [hunk 2]
```diff
@@ -0,0 +1,601 @@
+from ..utils import extract_attributes, float_or_none, int_or_none, parse_iso8601
+from .common import InfoExtractor
+
+
+class RheinMainTVIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?rheinmaintv\.de/sendungen/(?:[a-z-]+/)*(?P<display_id>[a-z-]+)/vom-(?P<date>[0-9]{2}\.[0-9]{2}\.[0-9]{4})(?:/(?P<serial_number>[0-9]+))?'
+    _TESTS = [{
+        'url': 'https://www.rheinmaintv.de/sendungen/beitrag-video/formationsgemeinschaft-rhein-main-bei-den-deutschen-meisterschaften/vom-14.11.2022/',
+        # 'md5': ...,  # left out because checksum is changing
```
Comment (pukkandan) [inline yt_dlp/extractor/rheinmaintv.py:9]:
```suggestion
```
This is expected for fragmented formats

File: yt_dlp/extractor/rheinmaintv.py [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+            return dictionary
+
+        ism_manifest_url = source['src'] or get(json, 'embedUrl')  # attribute 'src' is mandatory
+        formats, subtitles = self._extract_ism_formats_and_subtitles(ism_manifest_url, video_id)
+
+        def removeprefix(string, prefix):
+            return string[len(prefix):] if string.startswith(prefix) else None
+
+        def extract_format(internet_media_type, media):  # subtype of media (aka type)
+            return removeprefix(internet_media_type, media + '/') if internet_media_type is not None else None
```
Comment (pukkandan) [inline yt_dlp/extractor/rheinmaintv.py:582]: pointless functions. See utils

File: yt_dlp/extractor/rheinmaintv.py [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+        headline = self._html_search_regex(r'<h1><span class="title">([^<]*)</span>', webpage, 'headline')
+
+        source, img = self._search_regex(r'(?s)(?P<source><source[^>]*>)(?P<img><img[^>]*>)', webpage, 'video', group=('source', 'img'))
+        source = extract_attributes(source)
+        img = extract_attributes(img)
+
+        # json = self._search_json_ld(webpage, video_id)  # uncomment to extract the useless 'contentUrl' (as 'url') instead of the essential 'embedUrl'
+        #                                                 # then use get(json, 'title') instead of get(json, 'name')
+        json = self._parse_json(self._search_regex(r'(?s)<script type="application/ld\+json">([^<]*)</script>', webpage, 'json'), video_id)
+        # json = None  # uncomment to go without ld+json altogether
```
Comment (pukkandan) [inline yt_dlp/extractor/rheinmaintv.py:566]: remove comments

File: yt_dlp/extractor/rheinmaintv.py [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+
+        headline = self._html_search_regex(r'<h1><span class="title">([^<]*)</span>', webpage, 'headline')
+
+        source, img = self._search_regex(r'(?s)(?P<source><source[^>]*>)(?P<img><img[^>]*>)', webpage, 'video', group=('source', 'img'))
+        source = extract_attributes(source)
+        img = extract_attributes(img)
+
+        # json = self._search_json_ld(webpage, video_id)  # uncomment to extract the useless 'contentUrl' (as 'url') instead of the essential 'embedUrl'
+        #                                                 # then use get(json, 'title') instead of get(json, 'name')
+        json = self._parse_json(self._search_regex(r'(?s)<script type="application/ld\+json">([^<]*)</script>', webpage, 'json'), video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/rheinmaintv.py:565]: * dont use `json` as var name
~~* Use `self._search_json`~~ (see comment about json/ld)

File: yt_dlp/extractor/rheinmaintv.py [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+            'upload_date': '20221115'
+        }
+    }]
+
+    def _real_extract(self, url):
+        mobj = self._match_valid_url(url)
+        display_id = mobj.group('display_id')
+        video_id = 'vom %s' % mobj.group('date')
+        if mobj.group('serial_number'):
+            video_id += ' (%s)' % mobj.group('serial_number')
```
Comment (pukkandan) [inline yt_dlp/extractor/rheinmaintv.py:554]: * can be simplified
* Use only `\w-` for id, not spaces/`()`

File: yt_dlp/extractor/rheinmaintv.py [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+            return removeprefix(internet_media_type, media + '/') if internet_media_type is not None else None
+
+        extension = extract_format(source.get('type'), 'video')  # use video format as filename extension (!?)
+        if extension:
+            for f in formats:
+                f['ext'] = extension
+
+        return {
+            'id': video_id,
+            'title': headline or img.get('title') or get(json, 'name') or self._og_search_title(webpage) or self._html_extract_title(webpage).removesuffix(' -'),
```
Comment (pukkandan) [inline yt_dlp/extractor/rheinmaintv.py:591]:
```suggestion
            'title': (self._html_search_regex(r'<h1><span class="title">([^<]*)</span>',
                                              webpage, 'headline', default=None)
                      or img.get('title') or json.get('name')
                      or self._og_search_title(webpage, default=None)
                      or remove_end(self._html_extract_title(webpage), ' -')),
```

File: yt_dlp/extractor/rheinmaintv.py [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+            'alt_title': img['alt'],  # attribute 'alt' is mandatory
```
Comment (pukkandan) [inline yt_dlp/extractor/rheinmaintv.py:592]: not for us

File: yt_dlp/extractor/rheinmaintv.py [hunk 9]
```diff
@@ -0,0 +1,601 @@
...
+            'description': get(json, 'description') or self._og_search_description(webpage),
+            'display_id': display_id,
+            'formats': formats,
+            'subtitles': subtitles,
+            'thumbnail': img['src'] or get(json, 'thumbnailUrl'),  # attribute 'src' is mandatory
```
Comment (pukkandan) [inline yt_dlp/extractor/rheinmaintv.py:597]: same

File: yt_dlp/extractor/rheinmaintv.py (hunk 5 above)
Comment (pukkandan) [inline yt_dlp/extractor/rheinmaintv.py:565]: Do this https://github.com/yt-dlp/yt-dlp/blob/8791e78cccd68db8161f06dc8567280e0d99a5e1/yt_dlp/extractor/cnn.py#L189-L191

File: yt_dlp/extractor/common.py [hunk 10]
```diff
@@ -1263,7 +1263,10 @@ def _html_search_regex(self, pattern, string, name, default=NO_DEFAULT, fatal=Tr
         """
         res = self._search_regex(pattern, string, name, default, fatal, flags, group)
         if res:
-            return clean_html(res).strip()
+            if isinstance(res, (list, tuple)):
+                return tuple(clean_html(r) for r in res)
+            else:
+                return clean_html(res)
```
Comment (pukkandan) [inline yt_dlp/extractor/common.py:1269]: since this is unrelated to the PR, I'll merge it separately

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/rheinmaintv.py (hunk 1 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py:10]: That is how I started. Why would you prefer a dummy test (see md5 sum below) over a real one?

What is the purpose of these test cases anyway? I would like to catch any important change of the website, and this includes a lot more than the page title.

File: yt_dlp/extractor/rheinmaintv.py (hunk 2 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: Fine.
You might consider improving the extractor template. Otherwise, everybody will have to figure that out over and over again.

File: yt_dlp/extractor/rheinmaintv.py (hunk 3 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: Any suggestions?

File: yt_dlp/extractor/rheinmaintv.py (hunk 4 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: Done.

File: yt_dlp/extractor/rheinmaintv.py (hunk 5 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: Done.

File: yt_dlp/extractor/rheinmaintv.py (hunk 7 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: I hate superfluous parentheses, especially if their purpose is to please a policy tool.

File: yt_dlp/extractor/rheinmaintv.py (hunk 5 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: Done.

File: yt_dlp/extractor/rheinmaintv.py (hunk 6 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: Good point.
If url == os.path.join(contentUrl, video_id) is desired, however, different changes are required.
Done.

File: yt_dlp/extractor/rheinmaintv.py (hunk 8 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: Strange preferences.
Done.

File: yt_dlp/extractor/rheinmaintv.py (hunk 9 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: same

File: yt_dlp/extractor/rheinmaintv.py (hunk 1 above)
Comment (Grub4K) [reply yt_dlp/extractor/rheinmaintv.py:10]: The changes are cought if the returned `info_dict` has changed values. The suggestion of removing all properties is to generate the correct test case, not for the final test, which must contain all values. If it returns the same data after a change it still works as expected and should not be of concern.

File: yt_dlp/extractor/rheinmaintv.py (hunk 3 above)
Comment (Grub4K) [reply yt_dlp/extractor/rheinmaintv.py]: Use `traverse_obj` instead. See the [contributing.md](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#mandatory-and-optional-metafields) for further info. For example `get(d, 'a', 'b')` can be written as `traverse_obj(d, ('a', 'b'))`.

File: yt_dlp/extractor/rheinmaintv.py (hunk 7 above)
Comment (Grub4K) [reply yt_dlp/extractor/rheinmaintv.py]: The purpose of this policy is not to please a policy tool, but rather to create more readable code. Long lines are unreadable and the convention for this is to break it into more lines.

yt-dlp already is very lenient with its coding style so the conventions that it imposes should not be hard to follow.

File: yt_dlp/extractor/rheinmaintv.py (hunk 3 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: traverse_obj is of no use here.

File: yt_dlp/extractor/rheinmaintv.py (hunk 7 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: The lines won't get any longer without the parentheses.

File: yt_dlp/extractor/rheinmaintv.py (hunk 1 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py:10]: So it seems that the test cases are well suited for the final test?!

File: yt_dlp/extractor/rheinmaintv.py (hunk 3 above)
Comment (Grub4K) [reply yt_dlp/extractor/rheinmaintv.py]: I am showing you replacements for future reference (`get` => `utils.traverse_obj`, `removeprefix` => `utils.remove_start`, `extract_format` => `utils.mimetype2ext`).

`self._extract_ism_formats_and_subtitles(ism_manifest_url, video_id)` should already handle setting `ext`, do you need to overwrite these? If so, that would be worth a comment. In that case you should use `mimetype2ext`.

File: yt_dlp/extractor/rheinmaintv.py (hunk 7 above)
Comment (Grub4K) [reply yt_dlp/extractor/rheinmaintv.py]: If you want to indent without additional parenthesis you can just do so:
```py
            'title':
                self._html_search_regex(r'<h1><span class="title">([^<]*)</span>',
                                        webpage, 'headline', default=None)
                or img.get('title') or json_ld.get('title')
                or self._og_search_title(webpage)
                or remove_end(self._html_extract_title(webpage), ' -'),
```
`str.removesuffix` is added with Python 3.9, so the `utils` function is used instead.

File: yt_dlp/extractor/rheinmaintv.py (hunk 3 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: OMG, I overlooked mimetype2ext despite its pretty straightforward name. Thank you!

This removes one concern about this part of the code. The other is, as you mention, that it blindly overwrites the default extension. Apparently, .ismv (for video, .isma for audio) is hardwired in self._extract_ism_formats_and_subtitles. Of course, the file could be converted afterwards or even by yt-dlp itself when asked to do so. But I'd prefer to give it the original (?) extension right away. It always seems to be .mp4 for this site.

File: yt_dlp/extractor/rheinmaintv.py (hunk 7 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: Ah, finally this is a layout which flake8 accepts. Thank you!

I hope that the page title will never have to be extracted that way, but fixed that too.

**Comment by barthelmannk** [comment]
BTW, what do you think about code like this (replacing the monolithic info_dict at the end)? It tries to merge the entire json_ld dict. It refuses to use an existing library function, adopts a somewhat strange layout, and burns a few cpu cycles with unnecessary og_searches.
```
        # from ..utils import merge_dicts  # doesn't look trustworthy
        def merge_dicts(*dicts):
            merged = {}
            for d in dicts:
                for k, v in d.items():
                    if v is not None and k not in merged:
                        merged[k] = v
            return merged

        return merge_dicts({
            'id': video_id,
            'title':
                self._html_search_regex(r'<h1><span class="title">([^<]*)</span>',
                                        webpage, 'headline', default=None)
                or img.get('title'),
            'alt_title': img.get('alt'),
            'display_id': display_id,
            'formats': formats,
            'subtitles': subtitles,
            'thumbnails': [{'url': img['src']}] if 'src' in img else None,
        }, json_ld, {
            'title':
                self._og_search_title(webpage)
                or remove_end(self._html_extract_title(webpage), ' -'),
            'description': self._og_search_description(webpage),
        })
```

**Comment by Grub4K** [comment]
First of all why does `utils.merge_dicts` not look trustworthy? Secondly the `or` chain makes it immediately clear where and which of the attributes are used AND evaluates lazily. In this case you need to split the title attribute into three distinct parts, making it harder to understand.

Do you want to do it that way for readability or? If so, I just want to add that I like the previous more both for readability and understandability. I don't think it's worth a change.

**Comment by barthelmannk** [comment]
That's what I wanted to hear, thanks. The idea was to pull in the complete json_ld dict in one go - see above.

utils.merge_dicts isn't trustworthy because of the extra `or isinstance(v, str) and merged[k] == ''` in the test. This has nothing to do with the purpose of this function and it is a very ugly hack at best. I would never rely on such a function, and it should not be part of any library. Apparently, you are much more fussy about other people's code. ;-)

**Comment by Grub4K** [comment]
This check is so that `merge_dicts({'a': ''}, {'a': 'a'})` results in `{'a': 'a'}`. We would want the value with content in it and we cannot ignore all falsy values since `0` is a valid value. This behavior is specifically tuned to extractor usecases like this one. How does this check fail your usecase?

File: yt_dlp/extractor/rheinmaintv.py [hunk 11]
```diff
@@ -0,0 +1,588 @@
...
+        # Work around the method self._json_ld (called by self._search_json_ld), which
+        # extracts the useless 'contentUrl' (as 'url') instead of the essential 'embedUrl'.
+        raw_json_ld = list(self._yield_json_ld(webpage, video_id))
+        json_ld = self._json_ld(raw_json_ld, video_id)
+
+        ism_manifest_url = (
+            source.get('src')
+            or next(json_ld.get('embedUrl') for json_ld in raw_json_ld if json_ld.get('@type') == 'VideoObject')
+        )
+        formats, subtitles = self._extract_ism_formats_and_subtitles(ism_manifest_url, video_id)
```
Comment (Grub4K) [inline yt_dlp/extractor/rheinmaintv.py:569]: Instead of setting `ext` on the info dict itself, set it on each format like you did before (just using `mimetype2ext` instead). This also removes the need for the change in `YoutubeDL.py` and the ignored `ext` key in the info dict.
```suggestion
        formats, subtitles = self._extract_ism_formats_and_subtitles(ism_manifest_url, video_id)
        # The extension returned by the function is inaccurate and needs to be overwritten
        ext = mimetype2ext(source.get('type'))
        if ext:
            for f in formats:
                f['ext'] = ext
```

File: yt_dlp/extractor/rheinmaintv.py [hunk 12]
```diff
@@ -0,0 +1,588 @@
+from ..utils import extract_attributes, mimetype2ext, remove_end
+from .common import InfoExtractor
```
Comment (Grub4K) [inline yt_dlp/extractor/rheinmaintv.py:2]: sort the imports

Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py]: Done.

File: yt_dlp/extractor/rheinmaintv.py (hunk 11 above)
Comment (barthelmannk) [reply yt_dlp/extractor/rheinmaintv.py:569]: Done.

File: yt_dlp/extractor/common.py [hunk 13]
```diff
@@ -1267,7 +1267,7 @@ def _html_search_regex(self, pattern, string, name, default=NO_DEFAULT, fatal=Tr
         elif res:
             return clean_html(res).strip()
         else:
-            return res
+            return clean_html(res)
```
[the remark this answers is not in the document]
Comment (barthelmannk) [reply yt_dlp/extractor/common.py]: Yes, sorry.

BTW, I think this is bad code and suggest
```
        if isinstance(res, (list, tuple)):
            return tuple(clean_html(r) for r in res)
        else:
            return clean_html(res)
```
instead.
* `..util.clean_html` ends with `return html.strip()`, so the `.strip()` is clearly superfluous (and possibly harmful if `clean_html` would return `None` for any reason in the future).
* The combination of `isinstance(res, (list, tuple))` and `return tuple(...)` seems to be common practice in `common.py`. Why expect a tuple and return a list?
* If `not res` then `res` is either `''` or `None`, and `clean_html(res)` works perfectly fine in this case.

[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/common.py]: I hadn't noticed that clean_html does .strip

File: yt_dlp/extractor/rheinmaintv.py (hunk 1 above)
Comment (pukkandan) [reply yt_dlp/extractor/rheinmaintv.py:10]: No, it is not - which is obviously why I suggested how to generate them!

File: yt_dlp/YoutubeDL.py [hunk 14]
```diff
@@ -2017,7 +2017,7 @@ def can_merge():
             else 'bestvideo*+bestaudio/best' if not compat
             else 'bestvideo+bestaudio/best')
 
-    def build_format_selector(self, format_spec):
+    def build_format_selector(self, format_spec, default_output_format=None):
```
Comment (bashonly) [inline yt_dlp/YoutubeDL.py:2020]: Why are you changing this again? I thought this matter had already been settled.

Have you checked every other extractor to make sure this change won't have unintended consequences?

Comment (barthelmannk) [reply yt_dlp/YoutubeDL.py:2020]: It seems to me that this is a far more satisfactory and robust approach than overwriting file extensions. The change need not be merged in if you don't like it, though.

I don't know what other extractors intend when including an extension which is silently discarded. Perhaps the result will be more to their liking?

Comment (Grub4K) [reply yt_dlp/YoutubeDL.py:2020]: How the formats dicts work is it sorts them and picks the most fitting one, updating the base info dict with the format keys. That is why it seems like the `ext` is getting discarded, bc the format dicts `ext` key will overwrite the one in the base dict. Formats should stand for themselves, otherwise sorting them relative to another does not work consistently.

Please do explain why setting the extension for each format is unsatisfactory. It is how the system was designed to work and it seems more robust to me. Currently all extractors will set the extension on the formats either explicitly or implicitly, for easier support of non homogeneous extension in the formats. I see no benefit in changing it like this instead? Is it for convenience?

There was only [one change](https://github.com/yt-dlp/yt-dlp/pull/5840/files/a500e20ade2ccf8bd227b72bc8dc61c62676e07a#r1056789604) pending to get this PR into a mergeable state before reverting the `YoutubeDL` commit, which was conforming the tests to our test runner. Since you keep introducing other, potentially more drastic changes, we need to revise these (anew?) and question those changes.

We are open to changes that benefit the core, which is why I am asking you again and again on how this change is beneficial over our suggested solutions (May it only be convenience). However please understand that we need to understand these changes to be able to merge them into the code. Why not create an issue first explaining the problem, then after the quirks have been ironed out you can open a new PR with the related core changes. Core PRs tend to take a long(er) time to be merged bc of the potential implications. That way this PR can be merged more quickly.

I hope this answered all your questions.

Comment (barthelmannk) [reply yt_dlp/YoutubeDL.py:2020]: **1st and 2nd paragraph:** It is not so simple (at least) for the ism format (as far as I could figure it out). In that case, two formats - one video and one audio - have to be chosen and "merged" into a container. Therefore, none of the formats alone stands for the result. The final extension can be chosen (within certain bounds) - see `--merge-output-format`. So far, the merging procedure is chosen even before the `info_dict` is considered (to catch potential errors, as the comment tells us). But it is possible to include that information in a second pass, adding an optional parameter to `build_format_selector`. I think it would be nice to set the output format according to what the site thinks it should be. It can still be overridden by the user.

Overwriting the (correct) extensions in the formats is a rather strange approach. It works for `mp4`, apparently because that can contain video as well as audio and because the actual `protocol` is specified separately. Subsequently, this also determines the output format for merge, as desired. Is this more than a coincidence?

The contents of the `info_dict` are `update`d not with the final (merged) format but with the video part of the result. Therefore, some entries for the test are different from the json output, see `-j`. In particular, the value for `ext` does not necessarily reflect the actual file extension. Is this a bug?

**4th paragraph:** I'm not at all surprised that you are reluctant to changes at a central component. Having that said, however, I think the changes are not that difficult to understand this time (as opposed to my first try).

It may indeed be better to cancel this PR, because I'm not really satisfied with the code right now and even less with what you want me to do. Let's wait until, if ever, someone else requests support for this site.

File: yt_dlp/extractor/rheinmaintv.py [hunk 15]
```diff
@@ -0,0 +1,865 @@
...
+                    'fourcc': 'H264',
+                    'language': 'und',
+                    'codec_private_data': '0000000167640028ACD940780227E5C044000003000400000300C83C60C6580000000168EBECB22C',
+                    'sampling_rate': None,
+                    'channels': 2,
+                    'bits_per_sample': 16,
+                    'nal_unit_length_field': 4
+                }
+            }],
+            'subtitles': {},
```
Comment (Lesmiscore) [inline yt_dlp/extractor/rheinmaintv.py:270]:
```suggestion
```

File: yt_dlp/extractor/rheinmaintv.py [hunk 16]
```diff
@@ -0,0 +1,865 @@
...
```
Comment (Lesmiscore) [inline yt_dlp/extractor/rheinmaintv.py:540]:
```suggestion
```

File: yt_dlp/extractor/rheinmaintv.py [hunk 17]
```diff
@@ -0,0 +1,865 @@
...
```
Comment (Lesmiscore) [inline yt_dlp/extractor/rheinmaintv.py:810]:
```suggestion
```

File: yt_dlp/YoutubeDL.py (hunk 14 above)
Comment (Lesmiscore) [reply yt_dlp/YoutubeDL.py:2020]: Do demonstrate and show your wish in a new issue, not here.

Extension is set for each formats because some services provide multiple formats for a video/audio. YouTube is an example.
Most of the time it's not overwritten since base info_dict don't have `ext` set.

What merged extension should it be for ISM? The code is setting to MP4 all the time. It's not used at all. So the merged format must be MOV while downloaded formats are all MP4 if the website thinks so? Ridiculous.



> The contents of the `info_dict` are `update`d not with the final (merged) format but with the video part of the result. Therefore, some entries for the test are different from the json output, see `-j`. In particular, the value for `ext` does not necessarily reflect the actual file extension. Is this a bug?

Test cases in extractors are meant to test the output from extractor itself. Not a bug at all

Comment (Lesmiscore) [reply yt_dlp/YoutubeDL.py:2020]: Do demonstrate and show your wish in a new issue, not here.

Extension is set for each formats because some services provide multiple formats for a video/audio. YouTube is an example.
Most of the time it's not overwritten since base info_dict don't have `ext` set.

What merged extension should it be for ISM? The code is setting to MP4 all the time. It's not used at all. So the merged format must be MOV while downloaded formats are all MP4 if the website thinks so? Ridiculous.



> The contents of the `info_dict` are `update`d not with the final (merged) format but with the video part of the result. Therefore, some entries for the test are different from the json output, see `-j`. In particular, the value for `ext` does not necessarily reflect the actual file extension. Is this a bug?

Test cases in extractors are meant to test the output from extractor itself. Not a bug at all

Comment (barthelmannk) [reply yt_dlp/YoutubeDL.py:2020]: **3rd paragraph:** The merged extension should be `mp4`. The possibility to specify it should *remove* the need to overwrite standard extensions in all the formats. Did you read my comment?

You are right that there aren't many alternatives for the container: `mp4`, `mov`, `flv`, `avi`. Each of them is suitable for the intermediate files, too.

**Last paragraph:** Test cases *do not* test the output of the extractor itself. They test a fairly processed outcome of its output. In fact, they see an intermediate result of the processing, somewhat before the final result, which can be seen in the JSON dump. Did you read my comment? Why do I need to explain the functionality of *your* application to you?

And please don't mess with my code. You can do whatever you like during and after a pull.

**Comment by bashonly** [comment]
Closed by 98cb1eda7a4cf67c96078980dbd63e6c06ad7f7c
